Initial binheap properties challenge for Americium.

sageserpent-open opened the pull request.

sageserpent-open commented:
@DRMacIver , @jlink Pardon the request out of the blue, but if you folks could do a quick sanity check on this PR, that would be great. Just dipping my toes into the water with this first challenge.

I've used SBT for the build as I'm hoping to mix both Java and Scala examples together, as that's one of Americium's goals. Gradle is feasible for this too, but I'm rusty with it, so went to my comfort zone which is SBT.

I've more or less lifted the Jqwik example verbatim, although I added some invariant checks to make sure the generation of cases was doing what it should be doing. I'm not sure if these are technically binary heaps according to being almost balanced, but my impression is that this was never part of the original brief - or perhaps I've missed something?

UPDATE: I've added an additional test with the correct form of the sort.

sageserpent-open commented on pbt-libraries/americium/src/test/java/challenges/binheap/Heap.java:
```diff
@@ -0,0 +1,43 @@
+        if (null != left) {
+            assert head <= left.head;
+
+            left.checkInvariant();
```
I should remove this, as `Heap` is an immutable type - so what's the point of recursively checking the invariant, it is already established by construction for the left and right children.

sageserpent-open replied:
Done.

sageserpent-open commented on pbt-libraries/americium/src/test/java/challenges/binheap/Heap.java:
```diff
@@ -0,0 +1,43 @@
+        if (null != right) {
+            assert head <= right.head;
+
+            right.checkInvariant();
```
Same here.

sageserpent-open replied:
Done.

sageserpent-open commented:
After consideration, I won’t have the time to pursue all of the challenges. I’ll cancel this pull request accordingly - thank you if anyone has had a look in the meantime.

jlink commented:
You don’t have to tackle all challenges to make it worthwhile. I haven’t looked at your code yet, but I was planning to.

sageserpent-open commented:
@jlink Thanks for your interest and encouragement. In that case, I can work through a few more challenges over the next week(s) as a background activity. 

If this yields fruit then I can raise a new PR or reopen this one - I just wanted to avoid keeping folks hanging around waiting for my output.

sageserpent-open commented:
@jlink I've reopened this pull request - actually, the title is now a misnomer as the `deletion` and `difference` challenges have been added in as well now.

Some notes:

1. I've added some copies of the existing challenge test to sanity check the tests against corrected SUTs for the `binheap` and `deletion` challenges.
2. In some cases there are Java and Scala duplicates of the same test.
3. Where it's useful, I've added additional tests to show how test case generation can be made more efficient, or just done that to start with. As long as this doesn't leverage insider knowledge of the bug being exhibited, I think it's fair to play to the PBT's strengths; _**although this leads to ...**_
4. In the case of the `difference` test, I've noted that I've had to tweak the frequency of the test cases to suit what I believe is a fundamental requirement of the test suite, and added commentary and an extra test to elaborate on this. I hope this isn't too controversial, if you think that analysis is incorrect, I would be interested in your thoughts....

I've left the SBT build field nested in the directory specific to Americium to avoid trampling on the Jqwik sources. I simply invoke the SBT build there directly, likewise I import that directory in isolation into IntelliJ to work on it.

I've added reports for the three challenges implemented so far.

sageserpent-open commented:
As an aside, this exercise was tremendously useful for exposing some bugs in the recently added JUnit5 integration. Nothing to do with shrinkage as such, rather just trying out some variations in the obvious use cases flushed out some teething troubles that are now fixed in Americium 1.4.2 . I also found a more serious bug in Americium exposed when writing very, very long running tests to realise the thought experiment in the `difference` challenge, this has also been fixed.

So a good experience, all in all.

jlink commented:
I assume you’re aware of it, but what’s definitely missing are the parts on README.MD and an overview about Americium.

sageserpent-open commented:
Actually, no - it had completely slipped my mind, too much focus on the challenges. 

Thanks for the heads-up - I’ll take a look at what’s in the other tool sections and add something similar this week.

jlink commented:
Just tell me when you think the PR is ready for merge.

sageserpent-open commented:
I've added some documentation to parallel what's there for Jqwik. For now, I'd say that's enough to submit for review, barring any other unintended omissions. I may circle back and do some more challenges after this PR, but it's worth getting this PR closed one way or another prior to embarking on further work...

jlink reviewed (commented):
The one thing where I'd require clarification is the usage of SBT: Where to get it or how to install it.

jlink commented on pbt-libraries/americium/reports/deletion.md:
```diff
@@ -0,0 +1,22 @@
+
+## Normalization
+
+Americium currently uses a fixed internal randomisation seed, so always yields the same shrinkage:
```
Just shrinking with a given random seed seems to partially defy the purpose of the shrinking challenge. I guess you should try to move towards non-fixed seeds in the long run.

sageserpent-open commented:
It belongs to the same family of JVM build tools as Maven, Gradle etc. Just as with those tools, you can download and install for various OSs - see here:    

https://www.scala-sbt.org/

Essentially we’re talking about a bunch of Java files and a Maven Central library dependency; a Maven POM can be provided if needs be as a lowest common denominator build.

Sent from my iPhone

> On 4 May 2022, at 10:26, Johannes Link ***@***.***> wrote:
> 
> ﻿
> @jlink commented on this pull request.
> 
> The one thing where I'd require clarification is the usage of SBT: Where to get it or how to install it.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you modified the open/close state.

sageserpent-open commented:
Yes, I’ve considered that, but wanted to keep things simple and have repeatable tests as opposed to seeds being taken from the external environment/ time / network address etc.

What I looked for was good shrinkage with the same deterministic behaviour - not having to cast around with varying seeds to flush out a good shrinkage.

I prefer the approach of pulling more cases via the DSL’s limits in order to explore more variety in the test cases.

If enough folks express an interest in having a configurable or ‘random’ seed though, then it can be done, possibly with a JVM property.

Sent from my iPhone

> On 4 May 2022, at 10:37, Johannes Link ***@***.***> wrote:
> 
> ﻿
> @jlink commented on this pull request.
> 
> In pbt-libraries/americium/reports/deletion.md:
> 
> > @@ -0,0 +1,22 @@
> +# Report for Americium Shrinking on "deletion"
> +
> +This report was generated with Americium 1.4.2
> +
> +## Normalization
> +
> +Americium currently uses a fixed internal randomisation seed, so always yields the same shrinkage:
> Just shrinking with a given random seed seems to partially defy the purpose of the shrinking challenge. I guess you should try to move towards non-fixed seeds in the long run.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you modified the open/close state.

jlink commented:
> It belongs to the same family of JVM build tools as Maven, Gradle etc. Just as with those tools, you can download and install for various OSs - see here:

Maybe you can shortly mention that in the framework description and provide a link to sbt's website.

jlink commented:
> What I looked for was good shrinkage with the same deterministic behaviour - not having to cast around with varying seeds to flush out a good shrinkage.

Your framework, your choice :-)
However, I think that you're robbing property-based testing from one of its strength: Exploring the vast input range through pseudo-randomized variation over time. It happens in my projects every other week that a property fails that hasn't failed before because a new constellation is discovered that had never been tried before.

That said: I'm happy to merge when you are.

sageserpent-open commented:
> > What I looked for was good shrinkage with the same deterministic behaviour - not having to cast around with varying seeds to flush out a good shrinkage.
> 
> Your framework, your choice :-) However, I think that you're robbing property-based testing from one of its strength: Exploring the vast input range through pseudo-randomized variation over time. It happens in my projects every other week that a property fails that hasn't failed before because a new constellation is discovered that had never been tried before.
> 
> That said: I'm happy to merge when you are.

You make a good point there - there is a trade off between sampling a very large number of cases and waiting a long time for a test run to complete, versus sampling fewer cases but having variation across test runs. There is an example of this I’ve seen when using Scalacheck on the Plutonium project, although usually more cases is enough to get a failure for most tests I’ve seen.

For me, I want repeatability in test suites that I run locally when working on a feature or bug, but I can see the appeal of varying seeds when doing CI builds.

Maybe it’s time to raise a new issue in the backlog...


Thanks for taking a look at this @jlink - I shall add a note about SBT and put in some kind of Maven support. I’ll ping you when it’s done. Once that’s in, please merge it in, thank you. I’m motivated to add a few more challenges later on, but would like to land this work as-is.

sageserpent-open commented:
Ok, I've added a note on getting SBT, plus a fallback POM for folks who are only interested in Java + Maven. @jlink would you merge at your convenience, please...

jlink commented:
@sageserpent-open Thanks for the effort and contribution!
